cli/src/commands/squash.rs

```diff
@@ -65,15 +64,19 @@ pub(crate) struct SquashArgs {
     /// description(s) of the source revision(s)
     #[arg(long, short, conflicts_with = "message_paragraphs")]
     use_destination_message: bool,
-    /// Interactively choose which parts to squash
+    /// Interactively choose which parts to squash (implies --keep-empty)
     #[arg(long, short)]
     interactive: bool,
     /// Specify diff editor to be used (implies --interactive)
     #[arg(long, value_name = "NAME")]
     tool: Option<String>,
-    /// Move only changes to these paths (instead of all paths)
+    /// Move only changes to these paths instead of all paths (implies
+    /// --keep-empty)
     #[arg(conflicts_with_all = ["interactive", "tool"], value_hint = clap::ValueHint::AnyPath)]
     paths: Vec<String>,
+    /// If true, then empty revisions will be kept.
```

Comment:
I would make this more detailed:

"Keep revisions even if all of they are completely emptied by the squash operation"

Reply:
Relatedely, I slightly prefer `--keep-emptied`.

Reply:
We could also add a TODO to explain the similarity with `jj restore` in the common case of moving changes into a parent in a linear history (and that the operations are different in other cases).

Reply:
In retrospect, `--no-abandon` seems like a better choice to me. We are no longer doing anything conditional on whether the change is empty or not (it's purely based on the input the user provides now), so `--keep-empty` / `--keep-emptied` probably isn't a great choice..

Reply:
It's a valid point that `--keep-emptied` does less, but I think that I ultimately disagree. As long as a regular `jj squash` with no options collapses two commits into one, that should be considered the "normal" behavior.

If the normal behavior was to keep both commits, I'd call the other behavior `--abandon-emptied`.

We could even have both `--keep-emptied` and `--abandon-emptied`, ultimately (instead of `--no-keep-emptied` or `--no-abandon-emptied`).

Another way of thinking about this: perhaps the main reason to have a `--no-blah` option should be to override a previously specified `--blah` option (in a config, say). This is most natural if `--no-blah` is the default behavior without any options. I just phrased this rule, so I may have missed some important exception, but it sounds good to me right now. :)

But I'll keep thinking about it, and perhaps others will chime in.

Reply:
Hmm, I'm on the fence about this. In my opinion:
* If we only have one flag, `--no-abandon` is a better name than `--keep-emptied`
  * `--keep-emptied` just adds confusion. Why are we saying this when whether we keep a commit is completely unrelated to whether it's empty?
* If we have two flags, `--keep-emptied` and `--abandon-emptied` is probably the better choice. They're clearly opposites.

I would probably be opposed to just implementing `--keep-emptied` right now (though I wouldn't be opposed to implementing both `--keep-emptied` and `--abandon-emptied`).

Reply:
Two caveats:
1) `jj squash .` only works from the root directory.
2) I'm also not sure how obvious it is to a user - I never even considered it myself until you brought it up.

I'm leaning towards  "implement both `--keep-emptied` and `--abandon-emptied`" (with the default being `--abandon-emptied`, and `--interactive` and paths both implying `--keep-emptied`).

Reply:
If it's not so obvious, I think implying the default behavior is rather confusing. There's a valid use case of `jj squash --interactive --abandon-emptied`, and user might wonder why he has to specify `--abandon-emptied` only if `--interactive`.

Reply:
Sorry, I think I was being ambiguous, which may have let to some confusion.

The thing that I'm not sure how  obvious it is to a user is the fact that you can use `jj squash .` to specify all files in the repository.

Reply:
Ok, thanks for clarification.

I'm still not sure if `jj squash path` is clear sign to not abandon empty. To me, empty path arguments is usually equivalent to `.` or `root:"."`. Some commands like `jj log .` have subtle difference, but it's not super obvious either.

Reply:
That was why my original proposal involved splitting it into two commands - `movediff` to work on diffs, and `squash` to work on commits. That would be my preference, but it got some pretty strong pushback, so I went with this instead.

Reply:
Can't we simply add `--keep-emptied`/`--no-abandon` flag which is always off by default? User can add `diffmove` alias to turn it on. I don't think that'll provide the best UX, but it's less confusing than implying the default from path arguments.

Reply:
Maybe `--never-abandon` is an OK name? It could be shortened to `--na`. Or perhaps we can come up with something better (I'd still prefer something without a negative in the name).

I still dislike `--no-abandon`, perhaps since it sounds like, without that option, something will definitely be abandoned. (this thought is definitely influenced by Matt's point above that I also link to below). But I could live with it if we can't agree on something else. 

In a world where we don't change the default behavior as Yuya suggested just above, I think `--keep-emptied` is also OK, but I don't insist on it.

In a world where we do change the default behavior, Matt's point from https://github.com/martinvonz/jj/pull/3830#discussion_r1626918110 gives me more pause.

(One other small argument against `--keep-emptied` is that @martinvonz pointed out on Discord that we might one day want something like `jj squash --keep`, that copies diffs instead of moving them. Though, perhaps that could be named `jj copydiff` or something.)

---------

From this discussion and https://github.com/martinvonz/jj/pull/3830#discussion_r1627391465, it sounds like changing the default behavior is in fact controversial. I kind of liked the idea in the other thread, but I'm getting cold feet after reading Martin and Yuya's points. Perhaps we should just add an option for now?

Reply:
I'm not sure I follow all the ramifications of this, so I suggest to look at the big picture in terms of "changes" and "commits" and see if this is a special case. I've heard that you want to treat all commits the same way. Does that apply to all changes also? Is this a special case of `squash` or a `split`?

IMO, "changes" are the focus, and if you are squashing repeatedly, the emptied one is not a change to keep. Maybe it should be more clear if the flag is about the commit or the change.
I agree that most options should not be negative, but then don't most binary options get one of each, in order to override a config setting?

Reply:
I've simplified this change down to only adding the flag, and not changing the default behaviour. I called it `--keep-emptied` for now, but I'm not particularly fussed on what we end up deciding on - I'll just go with whatever people want.

> Is this a special case of squash or a split

I don't think I would call it either. I'd call this an "amend" TBH. In my mind (and maybe other people don't think about this the same way), "amend" means "update an existing change", while squash means "2 changes become one".

> if you are squashing repeatedly, the emptied one is not a change to keep

While I'm sure this is true for your workflow, it's not an absolute truth of squashes. In general, if anyone ever wants to have a branch pointer pointing to a "working copy" commit, rather than a "committed" commit, it becomes a problem.

As an example of a use case where you might want that, when I work on a bunch of related, but not dependent commits, I may end up with a change graph looking like (this is simplified - in reality, I've had as many as 10 commits in parallel):
```mermaid
graph TD;
    foo-->server_change;
    foo-->client_change;
    server_change-->proto_change;
    client_change-->proto_change;
    proto_change-->main;
```

Here:
* `foo` is my "feature branch". Similar to any other squash-based workflow, it's roughly speaking my "working copy".
* `proto_change` is sent out for review with no dependencies
* `client_change` and `server_change` are sent out for review with `proto_change` as a diffbase

My workflow is:
1) Someone asks for changes on `client_change`
1) I write those changes in `foo`, and verify that the changes actually work with the server changes
1) I run `jj amend --to client_change` - currently, this will move my feature branch `foo` to `client_change`, which is clearly incorrect. 

This might seem like a weird workflow for someone working on github, given that github tends to think in terms of a PR and branches (which I believe is just a sequential stack of changes), but for someone working on gerrit, where you send individual changes for review, this merge commit approach allows you to parallelize your code reviews, and is what I believe to be the natural approach.

While I'm fully aware that there are ways to work around this, why should I need to work around it? Wanting to have a branch or a topic associated with a commit that may, at times, be empty, isn't invalid, even if it's not a thing everyone wants.

Reply:
Thank you!

You currently called it `--keep-empty` (or perhaps it got renamed very recently). That's fine by me; I'm fine with either that or `--keep-emptied`.